src/pynxtools/dataconverter/validation.py

```diff
@@ -268,6 +183,854 @@ def is_valid_unit_for_node(
+                ''
+            """
+            if "/" not in path.strip("/"):
+                return ""
+            return path.rstrip("/").rsplit("/", 1)[0]
+
+        if _get_parent_path(variadic_name) == _get_parent_path(path):
+            node = find_node_for(variadic_name, node_type=node_type)
+            if node is not None and node.variadic:
+                score = get_nx_namefit(
+                    path.rsplit("/", 1)[-1],
+                    node.name,
+                    node.name_type == "any",
+                    node.name_type == "partial",
+                )
+                if score > -1:
+                    return True
+
+            return False
+
+    def remove_from_req_groups(path: str):
+        """
+        Remove a path from the set of required groups, accounting for variadic nodes.
+
+        Args:
+            path (str): The path to remove.
+        """
+        if path in required_groups:
+            required_groups.remove(path)
+        else:
+            # Check if a variadic required group exists
+            for grp in list(required_groups):
+                if _variadic_node_exists_for(path, grp, node_type="group"):
+                    required_groups.remove(grp)
+
+    def remove_from_req_entities(path: str):
+        """
+        Remove a path from the set of required entities (fields/attributes),
+        accounting for variadic nodes.
+
+        Args:
+            path (str): The path to remove.
+        """
+        if path in required_entities:
+            required_entities.remove(path)
+        else:
+            # Check if a variadic required node exists
+            for ent in list(required_entities):
+                node_type = "attribute" if "@" in ent else "field"
+
+                clean_path = (
+                    path.rstrip("/@units")
+                    if path.endswith("@units") and ent.endswith("@units")
+                    else path
+                )
+                clean_ent = (
+                    ent.rstrip("/@units")
+                    if path.endswith("@units") and ent.endswith("@units")
+                    else ent
+                )
+
+                if _variadic_node_exists_for(
+                    clean_path, clean_ent, node_type=node_type
+                ):
+                    required_entities.remove(ent)
+
+    def _check_for_nxcollection_parent(node: NexusNode):
+        """
+        Check if the given node has a parent group of type NXcollection.
+
+        Args:
+            node (NexusNode): The node to check.
+
+        Returns:
+            bool: True if a parent NXcollection group exists.
+        """
+        parent = node.parent
+        while parent:
+            if parent.type == "group" and parent.nx_class == "NXcollection":
+                # Found a parent collection group
+                return True
+            parent = parent.parent
+
+        return False
+
+    def check_reserved_suffix(path: str, parent_data: h5py.Group):
```

Comment:
I don't understand why this shows up as new code, was this not there since a long time?

Reply:
Ah, it seems a lot of functions are there now in two flavors, once in validate_hdf... and once in validate_dict...
Does this not make the code very hard to maintain? Could we reuse common code lines?

Reply:
great